### DISCUSSION TITLE
Is it possible to make a PWA with Service Workers in Brython?

### DISCUSSION THREAD

**Discussion by ykonoclast** [brython-dev/brython#2180] [Q&A] [answered]
I am building a simple web application. I had wanted to replace Javascript for a long time and so far Brython has been a godsend.

But, since it is a progressive web app, I need to register a Service Worker. So far I can only do that in Javascript, is there a way to do that in Brython?

I have tried several approaches to no avail :

**1 : registering a Brython file using window.navigator.serviceWorker.register(src).** 
This does not work : MIME error as the browser needs a Javascript file.

**2 : register the Service Worker in javascript, but call Brython functions with the event handlers (onInstall etc.)**
Very tricky to do : Brython is usually not started in time when Javascript begins running. 
Even with normal javascript files (not service workers) and careful syncing, I usually get an error of the form "not a function" when trying to pass a Brython function to javascript code (using Window.callback = callback for instance).

**3 : communicate with the service worker from Brython code using messages** 
Failed because I can't find a way to make the Service Worker wait until Brython is available as Service Workers can't access Window or other document properties.

I was able to make a Brython PWA, but only at the cost of writing the Service Worker in Javascript, confining Python to UI.
Is there a way to write the Service Worker in Brython?

**Comment by PierreQuentel** [comment, marked as answer]
Thanks for raising this issue, and sorry for the late reply.

I'm afraid using service workers will not be possible with Brython. The specification makes it explicit that the service worker file must be in Javascript.

For web workers, it was possible to convert a Brython script into Javascript, then build a url from a blob containing the JS code and pass it to `Worker(url)`. For service workers, because of security reasons, the constraint is more strict, the url passed to `navigator.serviceWorker.register(url)` must use the protocol http or https, so I don't think there is a possibility to write the service worker in Python.

As you say, the equivalent of **`app.js`** or **`image-list.js`** in the [service worker demo](https://github.com/mdn/dom-examples/tree/main/service-worker/simple-service-worker) can be written in Python, but since the core script **`sw.js`** cannot, I prefer to make it clear that Brython does not support this technology.

**Reply by ykonoclast** [reply]
Merci Pierre pour cette réponse claire et complète et pour le super boulot sur Brython.
